### ISSUE TITLE
sphgrid.m: Wrong number of triangles

### ISSUE THREAD

**Issue by florianquintes** [StollLab/EasySpin#343]
When calculating the Delaunay triangles, an error occurs in the case of point groups with `openphi==True`. The number of triangles depends on the number of octants and nodes of the grid, whereby (nKnots-1)**2 triangles are obtained for one octant.

As can be seen for D2h, this is true for `openphi==False`. For the example C4h, however, too many triangles are obtained. Several errors lead to this:
1. for triangles near the pole, due to the symmetry conditions, two points of the triangle have the identical angles, so that a line is obtained instead of a triangle (e.g. triangle [1 2 2]). These triangles are then sorted out on the basis of their area (line 190-192), which means that these angles do not (fully) contribute to the spectrum.
2. since too many triangles are obtained despite the sorting out of "triangles" with an area of 0, there must still be triangles that are obtained more than once. I have not yet investigated this in more detail, for this you would have to look at the indices of all triangles obtained.

This bug should only have a minor effect, as the majority of triangles are correct and the proportion decreases as the grid size increases.

D2h:
````
Sys = struct('S',1/2,'g',[2 2 2]);
Sys.lw = 0.5;
Exp = struct('CenterSweep',[325 80],'mwFreq',9.5,'Harmonic',1);
Opt = struct('Verbosity',2);
Opt.GridSize = [10 0];
Opt.GridSymmetry = 'D2h'

[x,y1] = pepper(Sys,Exp,Opt);
````
Output:

> -absorption spectrum construction----------------------
>   interpolation off
>   triangle/segment projection
>   spectrum array size: 1x1024 
>   total **81 triangles** (55 orientations), 1 transitions

[D2h_matlab.pdf](https://github.com/user-attachments/files/16049441/D2h_matlab.pdf)

C4h
````
Sys = struct('S',1/2,'g',[2 2 2]);
Sys.lw = 0.5;
Exp = struct('CenterSweep',[325 80],'mwFreq',9.5,'Harmonic',1);
Opt = struct('Verbosity',2);
Opt.GridSize = [10 0];
Opt.GridSymmetry = 'C4h'

[x,y1] = pepper(Sys,Exp,Opt);
````
Output:

> -absorption spectrum construction----------------------
>   interpolation off
>   triangle/segment projection
>   spectrum array size: 1x1024 
>   total **85 triangles** (46 orientations), 1 transitions

[C4h.pdf](https://github.com/user-attachments/files/16049433/C4h.pdf)

Both plots were obtained by adding the following code in sphgrid.m (line 190-194):
```
  rmv = Areas==0;
  Tri(rmv,:) = [];
  phx = [phi maxPhi*ones(1,GridSize-1)];           % added lines
  thx = [theta pi/2*(1:GridSize-1)/(GridSize-1)];  % copied from l. 139-142
  x = thx.*cos(phx);                               %
  y = thx.*sin(phx);                               %
  triplot(Tri, x, y)                               % Plot function
  Areas(rmv) = [];
  Tri = sort(Tri,2);
  Tri = uint32(Tri);
```

**Comment by stestoll** [comment]
I think there might be a slight misunderstanding how `sphgrid` etc work.

Under both D2h and C4h, the unique solid angle is one octant (phi = 0...pi/2, theta=0...pi/2). Therefore, the number of triangles for D2h and C4h, and their areas, are identical. They need to cover the the same solid angle after all. Code example:

```matlab
[gridC4h,triC4h] = sphgrid('C4h',4)
[gridD2h,triD2h] = sphgrid('D2h',4)

sum(triC4h.areas)/4/pi
sum(triD2h.areas)/4/pi

sort(triC4h.areas)-sort(triD2h.areas)
```

Now, for C4h, all orientations with phi=pi/2 give the same resonance fields as phi=0 and the same theta, so it is not necessary to calculate any orientations with phi=pi/2. Instead, one can use the resonance fields calculated orientations with phi=0. Therefore, there are fewer orientations in the C4h grid than in the D2h grid, even for the same number of triangles.

Spectra of a spin system with isotropic g value simulated using D2h and C4h grids are identical within numerical error:

```matlab

clear
Sys.g = 2;
Sys.lw = 0.5;
Exp.mwFreq = 9.5;
Exp.CenterSweep = [339.5 5];
Opt.GridSize = [10 0];

Opt.GridSymmetry = 'D2h';
[B,spcD2h] = pepper(Sys,Exp,Opt);
Opt.GridSymmetry = 'C4h';
[B,spcC4h] = pepper(Sys,Exp,Opt);

subplot(2,1,1)
plot(B,spcD2h,B,spcC4h);

subplot(2,1,2)
plot(B,spcD2h-spcC4h);
```

**Comment by florianquintes** [comment]
Hello,
Thank you for the detailed answer. I already know how sphgrid.m works and it is clearly described in your thesis.
I am also aware that the number of triangles must be identical and that is exactly my point. Hence I have also emphasised the output of easyspin in bold, where exactly this is not the case. Furthermore, a test example with an isotropic g value does not make sense, as the point group of the Hamilton operator is `O3` and therefore there is no angle dependence. This means that the spectra can only be identical.

To check whether `Opt.GridSymmetry = C4h` provides the correct result, you would need a system of the point group C4h and would have to simulate this once with `Opt.GridSymmetry = C4h` and once with e.g. `Opt.GridSymmetry = Ci` or `Opt.GridSymmetry = D2h`. Unfortunately, I was unable to find such a system right away.


The test 
```
[gridC4h,triC4h] = sphgrid('C4h',4)
[gridD2h,triD2h] = sphgrid('D2h',4)

sum(triC4h.areas)/4/pi
sum(triD2h.areas)/4/pi

sort(triC4h.areas)-sort(triD2h.areas)
```
you prepared also worked for me, but only in exactly this case. If the number of nodes is increased, the error I described occurs again:
```
[gridC4h,triC4h] = sphgrid('C4h', 10)
[gridD2h,triD2h] = sphgrid('D2h', 10)

sum(triC4h.areas)/4/pi
sum(triD2h.areas)/4/pi

sort(triC4h.areas)-sort(triD2h.areas)  % 85 != 81
```

**Comment by stestoll** [comment]
Thanks for the explanation, now I understand. You are absolutely right, this is a bug in `sphgrid`. Thanks for reporting!

The bug is a result of numerical errors during Delaunay triangulation using `delaunay` for open-phi grids for one and more octants. In these cases, `delaunay` sometimes returns spurious triangles with zero or near-zero areas, though not for small grid sizes.

This is now fixed with commit 67a5adb6d1b4f1b69b16caa3dfb9f30b4f907e84, where after triangulation any triangle with an area much smaller than the average triangle area is thrown out.

A cleaner solution would be to hard-code the triangulations for all grid types, similar to what is currently implemented for the important 1-octant closed-phi grid.
